runtime/compiler/infra/J9Cfg.cpp

```diff
@@ -1763,3 +1778,735 @@ J9::CFG::emitVerbosePseudoRandomFrequencies()
+ * nodes to get the last bytecode index for a given block, and from that get the bytecode.
+ * Since this will be done no less than 3 times, a helper function is provided.
+ */
+TR_J9ByteCode
+J9::CFG::getBytecodeFromIndex(int32_t index)
+   {
+   TR_ResolvedJ9Method *method = static_cast<TR_ResolvedJ9Method*>(_method->getResolvedMethod());
+   TR_J9VMBase *fe = _compilation->fej9();
+   TR_J9ByteCodeIterator bcIterator(_method, method, fe, _compilation);
+   bcIterator.setIndex(index);
+   return bcIterator.next();
+   }
+
+TR_J9ByteCode
+J9::CFG::getLastRealBytecodeOfBlock(TR::CFGNode *start)
+   {
+   return this->getBytecodeFromIndex(start->asBlock()->getExit()->getNode()->getLocalIndex());
+   }
+
+void
+J9::CFG::getInterpreterProfilerBranchCountersOnDoubletonMicroJIT(TR::CFGNode *cfgNode, int32_t *taken, int32_t *nottaken)
+   {
+   // We have not Real Tree Tops, so we need to work around that.
+   TR::Node *node = cfgNode->asBlock()->getExit()->getNode();
+
+   if (this != comp()->getFlowGraph())
+      {
+      TR::TreeTop *entry = (cfgNode->asBlock()->getNextBlock()) ? cfgNode->asBlock()->getNextBlock()->getEntry() : NULL;
+      _externalProfiler->getBranchCounters(node, entry, taken, nottaken, comp());
+      }
+   else
+      {
+      getBranchCounters(node, cfgNode->asBlock(), taken, nottaken, comp());
+      }
+
+   if (*taken || *nottaken)
+      {
+      if (comp()->getOption(TR_TraceBFGeneration))
+         dumpOptDetails(comp(), "If on node %p has branch counts: taken=%d, not taken=%d\n", node, *taken, *nottaken);
+      }
+   else if (isVirtualGuard(node))
+      {
+      *taken = 0;
+      *nottaken = AVG_FREQ;
+      int32_t sumFreq = summarizeFrequencyFromPredecessors(cfgNode, self());
+
+      if (sumFreq>0)
+         *nottaken = sumFreq;
+
+      if (comp()->getOption(TR_TraceBFGeneration))
+         dumpOptDetails(comp(), "Guard on node %p has default branch counts: taken=%d, not taken=%d\n", node, *taken, *nottaken);
+      }
+   else if (!cfgNode->asBlock()->isCold())
+      {
+      if (node->getBranchDestination()->getNode()->getBlock()->isCold())
+         *taken = 0;
+      else
+         *taken = LOW_FREQ;
+
+      if (cfgNode->asBlock()->getNextBlock() &&
+          cfgNode->asBlock()->getNextBlock()->isCold())
+         *nottaken = 0;
+      else
+         *nottaken = LOW_FREQ;
+
+      /*int32_t sumFreq = summarizeFrequencyFromPredecessors(cfgNode, this);
+
+      if (sumFreq>0)
+         {
+         *nottaken = sumFreq>>1;
+         *taken = *nottaken;
+         }
+      else
+         {
+         if (node->getByteCodeIndex()==0)
+            {
+            int32_t callCount = getParentCallCount(this, node);
+
+            // we don't know what the call count is, assign some
+            // moderate frequency
+            if (callCount<=0)
+               callCount = AVG_FREQ;
+
+            *taken = 0;
+            *nottaken = callCount;
+            }
+         }
+      */
```

Comment:
Been a while since we've been in this section of the code. If I remember correctly it is because the functions called in this commented section assume a valid TR CFG, while MJIT uses a CFG that looks more like the code size estimation CFG. It was commented in case we needed to implement this portion of the function later as far as I remember.

Reply:
To add on to Scott, this is basically a copy and paste from here https://github.com/eclipse-openj9/openj9/blob/9b82fccf82647edbb09c28e8b9db5a6ae34b74e5/runtime/compiler/infra/J9Cfg.cpp#L1452

Reply:
>this is basically a copy and paste from here

There's unfortunately a lot of existing code that needs to be cleaned up; however, we want to be critical about the quality of new code so that we're not increasing the amount of code that has to eventually be cleaned up. As such, just because the source did things in a dubious way doesn't mean it should be repeated elsewhere unless there's a very good reason to do so.

Reply:
okay, we'll take care in future.
For the mentioned comment, we did changes to the MJIT code and also cleaned up other surrounding code in the file; will push the changes to the PR for you to look if it still needs raising up an issue, thanks!